File: src/SWEA1861/Solution.js
/**
 * @param {number[][]} room
 * @return {number[]} [방 번호, 이동 횟수]
 */
const Solution = (room) => {
  
  const moveCount = new Array(room.length);
  for (var i = 0; i < moveCount.length; i++) {
    moveCount[i] = new Array(room.length).fill(0);
  }

  const answer = [0, 0];

  for (var row = 0; row < room.length; row++) {
    for (var col = 0; col < room.length; col++) {
      getMoveCount(room, moveCount, row, col);
      
      if (answer[1] < moveCount[row][col]) {
        answer[0] = room[row][col];
        answer[1] = moveCount[row][col];
      } else if (answer[1] == moveCount[row][col]) {
        answer[0] = Math.min(answer[0], room[row][col]);
      }
    }
  }

  return answer;
};

const getMoveCount = (room, moveCount, curRow, curCol) => {
  if (moveCount[curRow][curCol] !== 0) {
    // 이미 이전에 moveCount가 계산된 곳이면?
    // 불필요하게 다시 계산하지 않고 재귀를 종료함
    return;
  }
  
  const dxy = [
    [1, 0], [-1, 0], [0, 1], [0, -1],
  ];

  // 일단 현재 방을 방문했으니 1로 만든다
  moveCount[curRow][curCol] = 1;
  for (const each of dxy) {
    const nextRow = curRow + each[0];
    const nextCol = curCol + each[1];
    
    if (nextRow >= 0 && nextRow < room.length && nextCol >= 0 && nextCol < room.length) {
      // 방을 벗어나지 않는 범위에서
      if (room[nextRow][nextCol] == room[curRow][curCol] + 1) {
        // 정확히 1만큼 방 번호가 큰 곳이 발견됐다?

        // 해당 방에서 더 이동할 곳이 있는지 확인
        getMoveCount(room, moveCount, nextRow, nextCol);

        // 재귀를 한 후에 누적된 [nextRow][nextCol] 위치의 moveCount 값을 현재 위치의 moveCount와 합산
        moveCount[curRow][curCol] += moveCount[nextRow][nextCol];
      }
    }
  }

  // 4방향을 모두 탐색하고, 더 이상 이동할 좌표가 없으면 이 곳으로 와서 종료됨
};

// 3 2
console.log(
  Solution([
    [13, 2, 14, 12],
    [10, 16, 6, 5], 
    [1, 8, 3, 15], 
    [7, 11, 4, 9]
  ]
));

// 1 2
console.log(
  Solution([
    [1, 3],
    [2, 4]
  ]
));

// 1 4
console.log(
  Solution([
    [2, 3],
    [1, 4]
  ]
));

File: src/SWEA1861/Solution.java
package SWEA1861;

import java.io.BufferedReader;
import java.io.BufferedWriter;
import java.io.IOException;
import java.io.InputStreamReader;
import java.io.OutputStreamWriter;

public class Solution {

  static int roomSize;
  static int[][] room;
  static int[][] roomMoveCount;
  private static int[][] dXY = new int[][] {
    {-1, 0}, {1, 0}, {0, -1}, {0, 1}, // 상하좌우
  };

  public static void main(String[] args) throws IOException {
    BufferedReader bReader = new BufferedReader(new InputStreamReader(System.in));
    BufferedWriter bWriter = new BufferedWriter(new OutputStreamWriter(System.out));
    
    int testCaseCount = Integer.parseInt(bReader.readLine());

    for (int testCase = 1; testCase <= testCaseCount; testCase++) {
      // 새 테스트케이스 들어오면 정답 초기화
      int answerRoomNo = 0;
      int answerCount = 0;

      roomSize = Integer.parseInt(bReader.readLine());

      // 방 배열 초기화
      room = new int[roomSize][roomSize];
      roomMoveCount = new int[roomSize][roomSize];
      for (int i = 0; i < roomSize; i++) {
        String[] eachRow = bReader.readLine().split(" ");
        for (int j = 0; j < roomSize; j++) {
          room[i][j] = Integer.parseInt(eachRow[j]);
        }
      }

      // 각 위치를 방문하여 계산
      for (int i = 0; i < roomSize; i++) {
        for (int j = 0; j < roomSize; j++) {
          repeat(i, j);
          
          if (roomMoveCount[i][j] > answerCount) {
            answerCount = roomMoveCount[i][j];
            answerRoomNo = room[i][j];
          } else if (roomMoveCount[i][j] == answerCount) {
            answerRoomNo = Math.min(answerRoomNo, room[i][j]);
          }

        }
      }

      bWriter.write(String.format("#%d %d %d\n", testCase, answerRoomNo, answerCount));

    }
    
    bWriter.flush();

    bReader.close();
    bWriter.close();
  }

  private static void repeat(int startRow, int startCol) {
    if (roomMoveCount[startRow][startCol] > 0) {
      // 이미 이전에 방문해서, moveCount가 기록된 곳임
      return; // 또 계산할 필요 없음
    }

    roomMoveCount[startRow][startCol] = 1; // 기본적으로 처음 방문하면 1임
    
    for (int i = 0; i < dXY.length; i++) {
      int nextRow = startRow + dXY[i][0];
      int nextCol = startCol + dXY[i][1];

      if (nextRow >= 0 && nextRow < roomSize && nextCol >= 0 && nextCol < roomSize) {
        // 범위를 넘지 않는 4방에 대해서

        if (room[startRow][startCol] + 1 == room[nextRow][nextCol]) {
          // 정확히 1만큼 더한 방 번호가 발견된다면?
          
          repeat(nextRow, nextCol); // 다음 좌표도 이어진 숫자가 있나 확인

          roomMoveCount[startRow][startCol] = roomMoveCount[nextRow][nextCol] + 1;
        }

      }

    }

  }

}
